### PR TITLE
feat: asset findings relationship

### PR DESCRIPTION
**Pull request by paralta**
## Description

https://github.com/openclarity/vmclarity/blob/main/rfc/change-asset-findings-relationships.md

Further UI changes to be addressed in dedicated PR (e.g. asset details for a particular finding)

## Type of Change

[ ] Bug Fix  
[x] New Feature  
[x] Breaking Change  
[ ] Refactor  
[ ] Documentation  
[ ] Other (please describe)  

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/openclarity/vmclarity/blob/main/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [x] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

### CODE REVIEW

File: scanner/findingkey/vulnerability.go [hunk 1]
```diff
@@ -38,6 +38,14 @@ func (k VulnerabilityKey) VulnerabilityString() string {
 	return k.VulnerabilityName
 }
 
+// Filter returns a string that can be used to filter the vulnerability finding in the database.
+func (k VulnerabilityKey) Filter() string {
+	return fmt.Sprintf(
+		"findingInfo/vulnerabilityName eq '%s' and findingInfo/package/name eq '%s' and findingInfo/package/name eq '%s'",
```
Comment (akijakya) [inline scanner/findingkey/vulnerability.go:44]:
```suggestion
		"findingInfo/vulnerabilityName eq '%s' and findingInfo/package/name eq '%s' and findingInfo/package/version eq '%s'",
```

File: uibackend/server/dashboard_findings_impact.go [hunk 2]
```diff
@@ -477,8 +470,28 @@ func (s *ServerImpl) getFindingToAssetCountMapWithFilter(ctx context.Context, fi
 
 		findings := *f.Items
 		log.Debugf("Got findings %+v", findings)
-		if err = processFindings(findings, findingAssetMap, findingToAssetCount); err != nil {
-			return nil, fmt.Errorf("failed to process findings: %w", err)
+
+		for idx, finding := range findings {
+			assetCount := 0
```
Comment (akijakya) [inline uibackend/server/dashboard_findings_impact.go:475]: When I was testing I only saw zeros in the `Affected Assets` column of the Riskiest Assets widget, is this a UI related issue or the actual count should have been visible there?

File: ui/src/layout/Findings/utils.jsx [hunk 3]
```diff
@@ -1,44 +1,22 @@
...
-    },
-    {
-        Header: "Found on",
-        id: "foundOn",
-        sortIds: ["foundOn"],
-        accessor: original => formatDate(original.foundOn)
+        Header: "Last seen by",
+        id: "lastSeenBy",
+        sortIds: ["lastSeenBy"],
+        accessor: original => formatDate(original.lastSeenBy)
```
Comment (akijakya) [inline ui/src/layout/Findings/utils.jsx:10]: I think it shouldn't be formatted as `date`, as it is the id/name of the `assetScan`, right?

File: ui/src/layout/Findings/Secrets/TabSecretDetails.jsx [hunk 4]
```diff
@@ -22,7 +22,7 @@ const TabSecretDetails = ({data}) => {
                     <TitleValueDisplayRow>
                         <TitleValueDisplay title="File path">{filePath}</TitleValueDisplay>
                     </TitleValueDisplayRow>
-                    <FindingsDetailsCommonFields foundOn={foundOn} invalidatedOn={invalidatedOn} />
+                    <FindingsDetailsCommonFields firstSeen={firstSeen} lastSeen={lastSeen} />
```
Comment (akijakya) [inline ui/src/layout/Findings/Secrets/TabSecretDetails.jsx:25]: I'm not very great at UI but maybe the `LatSeenBy` should also be passed here?

File: api/server/database/gorm/database.go [hunk 5]
```diff
@@ -127,32 +128,9 @@ func initDataBase(config types.DBConfig) (*gorm.DB, error) {
 		return nil, fmt.Errorf("failed to create index providers_id_idx: %w", idb.Error)
 	}
 
-	// For processing asset scans to findings we need to find all the scan
```
[the remark this answers is not in the document]
Comment (paralta) [reply api/server/database/gorm/database.go:130]: updated, this index creation should have not been removed! thanks for catching this

File: ui/src/layout/Findings/utils.jsx (hunk 3 above)
Comment (paralta) [reply ui/src/layout/Findings/utils.jsx]: let's leave the UI changes for another PR, this PR only includes UI changes that will allow UI testing like removing fields that are no longer available

File: ui/src/layout/Findings/Secrets/TabSecretDetails.jsx (hunk 4 above)
Comment (paralta) [reply ui/src/layout/Findings/Secrets/TabSecretDetails.jsx:25]: https://github.com/openclarity/vmclarity/pull/1758#discussion_r1629625393

File: uibackend/server/dashboard_findings_impact.go (hunk 2 above)
Comment (paralta) [reply uibackend/server/dashboard_findings_impact.go:475]: looking into this now 👀

Comment (akijakya) [reply uibackend/server/dashboard_findings_impact.go:475]: Nice, it seems to be fixed now! 🔥

**Review by akijakya — APPROVED** [review]
LGTM, and I hope the UI will follow these changes soon :)

**Comment by paralta** [comment]
Blocked by https://github.com/openclarity/vmclarity/issues/1766

File: api/server/database/gorm/finding.go [hunk 6]
```diff
@@ -170,36 +207,124 @@ func (s *FindingsTableHandler) UpdateFinding(finding types.Finding) (types.Findi
...
+		}
+	}
+
+	// Construct filter based on discriminator type
+	// Use info properties that make the finding unique, check package scanner/findingkey.
+	var filter string
+	switch info := discriminator.(type) {
+	case types.PackageFindingInfo:
+		key := findingkey.GeneratePackageKey(info)
+		filter = fmt.Sprintf("id ne '%s' and ", *finding.Id) + key.Filter()
```
Comment (ramizpolic) [inline api/server/database/gorm/finding.go:279]: nit: filter seems to be the same for all cases and can be defined only once after switch

File: api/server/database/gorm/odata.go [hunk 7]
```diff
@@ -848,19 +848,15 @@ var schemaMetas = map[string]odatasql.SchemaMeta{
 	"Finding": {
```
Comment (ramizpolic) [inline api/server/database/gorm/odata.go:848]:
```suggestion
	findingSchemaName: {
```
can also be used for relationships against "Finding" table

**Review by ramizpolic — DISMISSED** [review]
well done @paralta, awesome work! 💯

**Comment by ramizpolic** [comment]
AFAIR we removed the test DB seed data, but if we still have that, we should also update it to use the new models

**Comment by paralta** [comment]
@ramizpolic thanks for reviewing! just updated with your comments.

https://github.com/openclarity/vmclarity/pull/1168 the demo data file was removed here

**Review by akijakya — APPROVED** [review]
🔥

**Review by akijakya — APPROVED** [review]

**Review by ramizpolic — APPROVED** [review]

**Pull request merged**